src/framework/components/anim/controller.js

```diff
@@ -4,17 +4,295 @@ import {
+}
+
+Object.defineProperties(AnimNode.prototype, {
+    parent: {
+        get: function () {
+            return this._parent;
+        }
+    },
+    name: {
+        get: function () {
+            return this._name;
+        }
+    },
+    path: {
+        get: function () {
+            return this._parent ? this._parent.path + '.' + this._name : this._name;
+        }
+    },
+    point: {
+        get: function () {
+            return this._point;
+        }
+    },
+    weight: {
+        get: function () {
+            return this._parent ? this._parent.weight * this._weight : this._weight;
+        },
+        set: function (value) {
+            this._weight = value;
+        }
+    },
+    speed: {
+        get: function () {
+            return this._speed;
+        }
+    },
+    animTrack: {
+        get: function () {
+            return this._animTrack;
+        },
+        set: function (value) {
+            this._animTrack = value;
+        }
+    }
+});
+
+function BlendTree(parent, name, point, type, parameters, children, findParameter) {
+    this._parent = parent;
+    this._name = name;
+    this._type = type;
+    this._parameters = parameters;
+    this._point = Array.isArray(point) ? new pc.Vec2(point) : point;
+    this._parameterValues = null;
+    this._weight = 1.0;
+    this._children = [];
+    this._findParameter = findParameter;
+    for (var i = 0; i < children.length; i++) {
+        var child = children[i];
+        if (child.children) {
+            this._children.push(new BlendTree(this, child.name, child.point, child.type, child.parameter ? [child.parameter] : child.parameters, child.children, findParameter));
+        } else {
+            this._children.push(new AnimNode(this, child.name, child.point, child.speed));
+        }
+    }
+}
+
+Object.defineProperties(BlendTree.prototype, {
+    parent: {
+        get: function () {
+            return this._parent;
+        }
+    },
+    name: {
+        get: function () {
+            return this._name;
+        }
+    },
+    point: {
+        get: function () {
+            return this._point;
+        }
+    },
+    weight: {
+        get: function () {
+            this.calculateWeights();
+            return this._parent ? this._parent.weight * this._weight : this._weight;
+        },
+        set: function (value) {
+            this._weight = value;
+        }
+    },
+    speed: {
+        get: function () {
+            return this._speed;
+        }
+    }
+});
+
+
+// Maths helper functions
+function between(num, a, b, inclusive) {
+    var min = Math.min(a, b),
+        max = Math.max(a, b);
+    return inclusive ? num >= min && num <= max : num > min && num < max;
+}
+
+function getAngleRad(a, b) {
+    return Math.atan2(a.x * b.y - a.y * b.x, a.x * b.x + a.y * b.y);
+}
+
+function clamp(num, min, max) {
+    return num <= min ? min : num >= max ? max : num;
+}
+
+Object.assign(BlendTree.prototype, AnimNode.prototype, {
```

Comment:
this approach to typing seems... unorthodox. Wouldn't it be better to have the usual base type with 'inheritance'?

Reply:
👍 I've moved this over to use es5s version of inheritance.